File: vb.net/GattConnections.vb
Public Class GattConnections : Inherits wclThread
    Private FAddress As Int64
    Private WithEvents FClient As wclGattClient
    Private FRadio As wclBluetoothRadio
    Private FLog As ListBox

    Private Delegate Sub UpdateLogDelegate(ByVal Msg As String)

    Private Sub UpdateLog(ByVal Msg As String)
        FLog.Items.Add(Msg)
    End Sub

    Private Sub Trace(ByVal Msg As String)
        Dim d As New UpdateLogDelegate(AddressOf UpdateLog)
        FLog.Invoke(d, FAddress.ToString("X12") + ":: " + Msg)
    End Sub

    Private Sub FClient_OnConnect(ByVal Sender As Object, ByVal [Error] As Integer) Handles FClient.OnConnect
        If [Error] <> wclErrors.WCL_E_SUCCESS Then
            Trace("Connection error: " + [Error].ToString("X8"))
        Else
            Trace("Client connected")

            Dim Services() As wclGattService = Nothing
            Dim Res As Integer = FClient.ReadServices(wclGattOperationFlag.goNone, Services)
            If Res <> wclErrors.WCL_E_SUCCESS Then
                Trace("Failed to read services: " + Res.ToString("X8"))
            Else
                Trace("Services have been read with success")
            End If
        End If
    End Sub

    Protected Overrides Function OnInitialize() As Boolean
        FClient = New wclGattClient()
        FClient.Address = FAddress
        Dim Res As Integer = FClient.Connect(FRadio)
        If Res <> wclErrors.WCL_E_SUCCESS Then
            Trace("Start connection failed: " + Res.ToString("X8"))
            FClient = Nothing
            Return False
        Else
            Return True
        End If
    End Function

    Protected Overrides Sub OnTerminate()
        If Not FClient Is Nothing Then
            FClient.Disconnect()
            FClient = Nothing
        End If
    End Sub

    Public Sub New(ByVal Radio As wclBluetoothRadio, ByVal Address As Int64, ByVal Log As ListBox)
        MyBase.New()

        FAddress = Address
        FRadio = Radio
        FLog = Log
    End Sub

    Public ReadOnly Property Address() As Int64
        Get
            Return FAddress
        End Get
    End Property
End Class

File: .NET/c#/GattConnections.cs
using System;
using System.Windows.Forms;

using wclCommon;
using wclBluetooth;

namespace ConcurrentBleConnections
{
    public class GattConnections : wclThread
    {
        private Int64 FAddress;
        private wclGattClient FClient;
        private wclBluetoothRadio FRadio;
        private ListBox FLog;

        private void Trace(String Msg)
        {
            Synchronize(FAddress.ToString("X12") + ":: " + Msg);
        }

        private void FClient_OnConnect(Object Sender, Int32 Error)
        {
            if (Error != wclErrors.WCL_E_SUCCESS)
                Trace("Connection error: " + Error.ToString("X8"));
            else
            {
                Trace("Client connected");

                wclGattService[] Services;
                Int32 Res = FClient.ReadServices(wclGattOperationFlag.goNone, out Services);
                if (Res != wclErrors.WCL_E_SUCCESS)
                    Trace("Failed to read services: " + Res.ToString("X8"));
                else
                    Trace("Services have been read with success");
            }
        }

        protected override Boolean OnInitialize()
        {
            FClient = new wclGattClient();
            FClient.OnConnect += FClient_OnConnect;
            FClient.Address = FAddress;
            Int32 Res = FClient.Connect(FRadio);
            if (Res != wclErrors.WCL_E_SUCCESS)
            {
                Trace("Start connection failed: " + Res.ToString("X8"));
                FClient = null;
                return false;
            }
            return true;
        }

        protected override void OnTerminate()
        {
            if (FClient != null)
            {
                FClient.Disconnect();
                FClient = null;
            }
        }

        protected override void OnSynchronize(object Param)
        {
            FLog.Items.Add(Param as String);
        }
        
        public GattConnections(wclBluetoothRadio Radio, Int64 Address, ListBox Log)
            : base()
        {
            FAddress = Address;
            FRadio = Radio;
            FLog = Log;
        }

        public Int64 Address { get { return FAddress; } }
    };
}
